app/js/S3Persistor.js

```diff
@@ -56,20 +67,73 @@ async function sendFile(bucketName, key, fsPath) {
   return sendStream(bucketName, key, readStream)
 }
 
-async function sendStream(bucketName, key, readStream) {
+async function sendStream(bucketName, key, readStream, sourceMd5) {
   try {
+    // if there is no supplied md5 hash, we calculate the hash as the data passes through
+    let hashPromise
+    let b64Hash
+
+    if (sourceMd5) {
+      b64Hash = hexToBase64(sourceMd5)
+    } else {
+      const hash = crypto.createHash('md5')
+      hash.setEncoding('hex')
+      pipeline(readStream, hash)
+      hashPromise = new Promise((resolve, reject) => {
+        readStream.on('end', () => {
+          hash.end()
+          resolve(hash.read())
+        })
+        readStream.on('error', err => {
+          reject(err)
+        })
+      })
+    }
```

Comment:
As for the other hash, I think this could be simplified to 

```
      const hash = crypto.createHash('md5').setEncoding('hex')
      hashPromise = pipeline(readStream, hash).then(() => {
        return hash.read()
      })
```